docs/wrapper-layer.md

```diff
@@ -0,0 +1,53 @@
+The OmnICU Wrapper Layer
+========================
+
+A key piece of rolling out ergonomic OmnICU APIs for Rust and other target languages is the *wrapper layer*: code that sits between the client and the lower-level OmnICU libraries.  This document explains the design behind the OmnICU wrapper layer.
+
+## Background
+
+### Ergonomic API versus Logical API
+
+When a programmer writes code using Intl, the code they write should demonstrate their *intent*.  They should use what I'm calling the *ergonomic API* of OmnICU.
+
+A *logical API*, on the other hand, is one that is closely tied to specific to functional units, like currencies or measurement units.  A logical API function should require a specific hunk of data and perform a specific task.
+
+Although ergonomic APIs correspond with what users of i18n intend to do, from an implementation standpoint, they are often quite heavy-weight, pulling in many dependencies that may not be required for all call sites.  Intl.NumberFormat is an example: it supports currency formatting, which requires a lot of code and data, but when formatting a number that's not a currency, you shouldn't need to carry all that extra code and data.
+
+The goal of the *wrapper layer* is to provide an ergonomic API for clients, but map it to the logical API under the hood.
+
+![Ergonomic to Logical API](assets/ergonomic-logical.svg)
+
+For example, in the above diagram, a client uses the ergonomic Intl.NumberFormat API with options to enable compact notation in long form.  The wrapper layer performs a bit of logic to figure out that this combination of options requires two functional units: plural rules and compact decimal format.  It then delegates to those two functions in the logical API.
+
+## Key Requirements for the Wrapper Layer
+
+### Host Language
+
+Since the ergonomic API depends on the host language (e.g., Objective-C calling into OmnICU via an FFI), the wrapper layer should be written in each individual host language, mapping to the logical API exported by the OmnICU library.
+
+This allows different languages to choose different styles for passing options into the library; for example, a Builder pattern could be used in Java, and an options bag could be used in JavaScript.
+
+### Code Slicing
+
+The wrapper layer should be suitable for tools to perform automatic dead code elimination.  By resolving options at compile time, the set of logical APIs required for a certain application can be known in time to perform static analysis; unused logical APIs can then be removed, reducing code and data size.
```

Comment:
This is for example critical to JS Closure that does all of the code elimination at compile time.